### PR TITLE
AB#627: Add enclave.json as payload to the executable

### PR DESCRIPTION
**Pull request by Nirusu**
This is still missing the required bits in Edgeless RT to be handled from within the application, however guess this can be reviewed by looking at the outside from it.

The test is a bit tedious and long, but I wanted it to be a bit more explicit so we do not break any binaries.

### CODE REVIEW

**Review by thomasten — CHANGES_REQUESTED** [review]

File: internal/cli/sign_test.go [hunk 1]
```diff
@@ -147,6 +150,156 @@ NumTCS=32
...
+	unsignedExeRealfs, err := hostFs.Open("../test/test_sign_elf_unsigned")
+	require.NoError(err)
+	unsignedExeRealfsStat, err := unsignedExeRealfs.Stat()
+	require.NoError(err)
+	unsignedExeRealfsSize := unsignedExeRealfsStat.Size()
+	copyBuffer := make([]byte, int(unsignedExeRealfsSize))
+	n, err := unsignedExeRealfs.Read(copyBuffer)
+	require.NoError(err)
+	require.EqualValues(unsignedExeRealfsSize, n)
+	err = fs.WriteFile("helloworld_unsigned", copyBuffer, 0644)
```
[the remark this answers is not in the document]
Comment (Nirusu) [reply internal/cli/sign_test.go:173]: They can be the same, that was just for a conceptual distinction between the two.

File: internal/cli/sign.go [hunk 2]
```diff
@@ -180,3 +193,94 @@ func (c *Cli) Sign(filename string) error {
...
+	defer f.Close()
+
+	// Check if a payload already exists
+	payloadSize, oeInfoOffset, err := checkIfPayloadExists(f)
+	if err != nil {
+		return err
+	}
+
+	// If a payload already exists, truncate the file to remove it
+	if payloadSize > 0 {
```
[the remark this answers is not in the document]
Comment (Nirusu) [reply internal/cli/sign.go:215]: Added those checks:

```golang
// Check if payload is at expected location
expectedPayloadSize := fileStat.Size() - int64(payloadSize)
if expectedPayloadSize != payloadOffset {
	return errors.New("expected payload location does not match real payload corruption, cannot safely truncate old payload")
}
```

```golang
else if (payloadSize == 0) != (payloadOffset == 0) {
	return errors.New("payload information in header is inconsistent, cannot continue")
}
```

Do you think they are sufficient?

File: internal/cli/sign_test.go [hunk 3]
```diff
@@ -147,6 +169,147 @@ NumTCS=32
...
+
+	// Offset should be the same as the unsigned executable
+	oeInfoPayloadOffsetUint64 := binary.LittleEndian.Uint64(oeInfoPayloadOffset)
+	assert.EqualValues(unsignedExeMemfsSize, oeInfoPayloadOffsetUint64)
+
+	// Reconstruct the JSON and check equality
+	reconstructedJSON := make([]byte, int(oeInfoPayloadSizeUint64))
+	n, err = signedExeMemfs.ReadAt(reconstructedJSON, int64(oeInfoPayloadOffsetUint64))
+	require.NoError(err)
+	require.EqualValues(oeInfoPayloadSizeUint64, n)
```
Comment (thomasten) [inline internal/cli/sign_test.go:257]:
```suggestion
	// Reconstruct the JSON and check equality
	reconstructedJSON := make([]byte, payloadSize)
	n, err := signedExeMemfs.ReadAt(reconstructedJSON, payloadOffset)
	require.NoError(err)
	require.EqualValues(payloadSize, n)
```

File: internal/cli/sign_test.go [hunk 4]
```diff
@@ -147,6 +169,147 @@ NumTCS=32
...
+	require.NoError(err)
+	expectedLengthOfPayload := len(jsonData)
+
+	// Embed json config to helloworld
+	err = cli.embedConfigAsPayload("helloworld", jsonData)
+	assert.NoError(err)
+
+	// Check if new helloworld_signed now contains signed data
+	signedExeMemfs, err := fs.Open("helloworld")
+	require.NoError(err)
```
Comment (thomasten) [inline internal/cli/sign_test.go:221]:
```suggestion
	require.NoError(err)
	defer signedExeMemfs.Close()
```
instead of cleanup at the end

**Review by thomasten — CHANGES_REQUESTED** [review]

File: internal/cli/sign_test.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (thomasten) [reply internal/cli/sign_test.go]: pls use a const for all occurrences of "helloworld"

**Review by thomasten — APPROVED** [review]

**Pull request merged**